compiler/rustc_metadata/src/locator.rs

```diff
@@ -1000,223 +998,159 @@ impl CrateError {
-                E0523,
-                "found two different crates with name `{}` that are not distinguished by differing \
-                 `-C metadata`. This will result in symbol conflicts between the two.",
-                root_name,
-            ),
             CrateError::StableCrateIdCollision(crate_name0, crate_name1) => {
-                let msg = format!(
-                    "found crates (`{}` and `{}`) with colliding StableCrateId values.",
-                    crate_name0, crate_name1
-                );
-                sess.struct_span_err(span, &msg)
+                sess.emit_err(StableCrateIdCollision {
+                    span,
+                    crate_name0: crate_name0.to_string(),
+                    crate_name1: crate_name1.to_string(),
+                });
+            }
+            CrateError::DlOpen(s) | CrateError::DlSym(s) => {
+                sess.emit_err(DlError { span, err: s.to_string() });
             }
-            CrateError::DlOpen(s) | CrateError::DlSym(s) => sess.struct_span_err(span, &s),
             CrateError::LocatorCombined(locator) => {
                 let crate_name = locator.crate_name;
-                let add = match &locator.root {
+                let add_info = match &locator.root {
                     None => String::new(),
                     Some(r) => format!(" which `{}` depends on", r.name),
                 };
-                let mut msg = "the following crate versions were found:".to_string();
-                let mut err = if !locator.crate_rejections.via_hash.is_empty() {
-                    let mut err = struct_span_err!(
-                        sess,
-                        span,
-                        E0460,
-                        "found possibly newer version of crate `{}`{}",
-                        crate_name,
-                        add,
-                    );
-                    err.note("perhaps that crate needs to be recompiled?");
+                // FIXME: Is there any way to get these notes and helps onto every diagnostic in this
+                // huge branch arm without changing them all to manual implementations?
+                let mut global_loc_notes = Vec::new();
+                let mut global_loc_helps = Vec::new();
+                if !locator.crate_rejections.via_filename.is_empty() {
+                    let mismatches = locator.crate_rejections.via_filename.iter();
+                    for CrateMismatch { path, .. } in mismatches {
+                        global_loc_notes.push(format!(
+                            "extern location for {} is of an unknown type: {}",
+                            crate_name,
+                            path.display(),
+                        ));
+                        global_loc_helps.push(format!(
+                            "file name should be lib*.rlib or {}*.{}",
+                            locator.dll_prefix, locator.dll_suffix
+                        ));
+                    }
```

Comment:
These "notes" and "helps" used to be (conditionally) added to the `err` way down at the end of the branch arm (see red lines 1195-1208 below).

I moved them up to the top and shoved the note/help strings into some vectors because I anticipated the need to add the strings as notes/helps to each of the new error types in this branch arm somehow, as I explained in my `FIXME` above.

BUT everything passes tests/CI!!!

??? So, is this:
a. _really_ dead code code that we should remove? (🤞🏻)
b. a gap in testing that we need to add tests for and then find a way to make work here? (😬)

I pushed up a commit where I inserted an `assert!(false);` to line `1046` and re-ran `x test` to find out if a larger range of tests ever hit this code. If they do, we know it's _not_ dead code. Otherwise, we'll know that at least as far as tests are concerned, it's dead code and _probably_ no one ever saw these notes/helps. 🤞🏻

Reply:
CI passes with the `panic!()`. Should I remove this block of code?

Reply:
It's likely that this is just untested code - `rustc_metadata`'s errors aren't going to be that common, as they relate to issues loading/decoding the cross-crate metadata that rustc uses to know about your dependencies - if we're doing our job right, we shouldn't hit these, and I imagine that they could be quite difficult to write test cases for. I'd just assume that the errors are necessary for now.

Reply:
Okay--I'll get them into the error then, somehow.

Reply:
Oh! Since they're so rare and don't have any tests, I suppose there's nothing preventing me from creating new error(s) to emit, rather than tacking them onto all the umpteen existing errors. That should be much cleaner.

Reply:
✅ Done.